### DISCUSSION TITLE
Is it somehow possible (without deep hacking) to attach cocalc project to some existing OS user?

### DISCUSSION THREAD

**Discussion by belonesox** [sagemathinc/cocalc#6277] [Q&A]
By default, cocalc-hub creates ephemeral users when corresponding projects are running and deletes then after project is stopped.

Is it somehow possible (with deep hacking) to attach cocalc project to some existing OS user?  
User, that exists before and after the project is active.

**Comment by williamstein** [comment]
> Is it somehow possible (with deep hacking) to attach cocalc project to some existing OS user?

What is the context?  Cocalc-docker?  https://github.com/sagemathinc/cocalc-docker

This functionality doesn't exist right now.  It could be implemented.  Currently cocalc-docker uses this code to control projects:

https://github.com/sagemathinc/cocalc/blob/master/src/packages/server/projects/control/multi-user.ts

You would write a variant of that which, e.g., implements "createUser" differently:

https://github.com/sagemathinc/cocalc/blob/master/src/packages/server/projects/control/util.ts#L119

**Comment by belonesox** [comment]
>What is the context? Cocalc-docker? https://github.com/sagemathinc/cocalc-docker

No, no docker.

Usecase:  I just want to use cocalc-hub + it integration with code-server  for some build server to enable real time collaboration on build sources (little "sledge-hammer to crack a nut", I can install code-server, but why not — collaborative notebooks will be usable).  One project with hardcoded user and home directory will be enough for that. 

Also, selecting username for the project it can be usable to solve some license issues with proprietary software, that binded its license to username (GUROBI, COPT, etc).

>This functionality doesn't exist right now. ...  You would write a variant of that which, e.g., implements "createUser" differently...

Ok, thank you, I had to check. May be I will try to hack a little for this case.

**Reply by williamstein** [reply]
That's a really interesting use case, and cocalc with some modifications can solve it, but certainly doesn't today.    Please let me know what happens.    You might be able to just put in some hardcoded mapping from a project id to a unix account name in that code I linked to...
